src/promptflow/promptflow/_utils/exception_utils.py

```diff
@@ -255,6 +248,53 @@ def error_code_recursed(self):
 
         return current_error
 
+    def to_dict(self, *, include_debug_info=False):
+        """Return a dict representation of the exception.
+
+        This dict specification corresponds to the specification of the Microsoft API Guidelines:
+        https://github.com/microsoft/api-guidelines/blob/vNext/Guidelines.md#7102-error-condition-responses
+
+        Note that this dict representation the "error" field in the response body of the API.
+        The whole error response is then populated in another place outside of this class.
+        """
+        if isinstance(self._ex, JsonSerializedPromptflowException):
+            return self._ex.to_dict(include_debug_info=include_debug_info)
+
+        # Otherwise, return general dict representation of the exception.
+        result = {"message": str(self._ex), "messageFormat": "", "messageParameters": {}}
+        result.update(self.error_code_recursed)
+
+        if include_debug_info:
+            result["debugInfo"] = self.debug_info
+
+        return result
+
+
+class PromptflowExceptionPresenter(ExceptionPresenter):
+    @property
+    def error_codes(self):
+        """The hierarchy of the error codes.
+
+        We follow the "Microsoft REST API Guidelines" to define error codes in a hierarchy style.
+        See the below link for details:
+        https://github.com/microsoft/api-guidelines/blob/vNext/Guidelines.md#7102-error-condition-responses
+
+        For subclass of PromptflowException, use the ex.error_codes directly.
+
+        For PromptflowException (not a subclass), the ex.error_code is None.
+        The result should be:
+        ["SystemError", {inner_exception type name if exist}]
+        """
+        if self._ex.error_codes:
+            return self._ex.error_codes
+
+        # For PromptflowException (not a subclass), the ex.error_code is None.
+        # Handle this case specifically.
+        error_codes = [infer_error_code_from_class(SystemErrorException)]
+        if self._ex.inner_exception:
+            error_codes.append(infer_error_code_from_class(self._ex.inner_exception.__class__))
```

Comment:
Is this part newly added or is it existing logic?

Reply:
The implementation has changed here, but the logic remains the same as before.
![image](https://github.com/microsoft/promptflow/assets/94442403/4d7f900b-9b1a-4ace-beba-87cd254f76ca)